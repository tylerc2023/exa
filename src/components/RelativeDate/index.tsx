import dayjs from 'dayjs';
import React, {memo, useEffect, useState} from 'react';
import relativeTime from 'dayjs/plugin/relativeTime';
import {Text} from '../Text';
import {TouchableOpacity} from '../TouchableOpacity';
dayjs.extend(relativeTime);

interface Props {
  date: number;
}

const minute = 60 * 1000;
export const RelativeDate = memo(function RelativeDate({date}: Props) {
  const [showRelativeDate, setShowRelativeDate] = useState(true);
  const [update, setUpdate] = useState(1);

  const toggleRelativeDate = () => {
    setShowRelativeDate((prev) => !prev);
  };

  useEffect(() => {
    const interval = setInterval(() => setUpdate((prev) => prev + 1), minute);
    return () => clearInterval(interval);
  }, []);

  return (
    <TouchableOpacity onPress={toggleRelativeDate}>
      {showRelativeDate && update ? (
        <Text title={dayjs(date).fromNow()} />
      ) : (
        <Text title={dayjs(date).format('MMM D YYYY, h:mm a')} />
      )}
    </TouchableOpacity>
  );
});
