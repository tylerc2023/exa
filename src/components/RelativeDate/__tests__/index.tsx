import * as React from "react";
import { create, ReactTestRenderer } from "react-test-renderer";
import RelativeDate from "..";

describe("RelativeDate component", () => {
  const date = Date.now();
  let dom: ReactTestRenderer;

  beforeEach(() => {
    dom = create(<RelativeDate date={date} />);
  });

  afterEach(() => {
    dom.unmount();
  });

  it("renders correctly", () => {
    const dom = create(<RelativeDate date={date} />).toJSON();
    expect(dom).toMatchSnapshot();
  });

  it("renders correctly on touch", () => {
    const dom = create(<RelativeDate date={1553904714478} />);
    dom.root.instance.toggleRelativeDate();
    expect(dom).toMatchSnapshot();
  });

  it("updates state touch", () => {
    const dom = create(<RelativeDate date={date} />);
    expect(dom.root.instance.state.showRelativeDate).toBe(true);
    dom.root.instance.toggleRelativeDate();
    expect(dom.root.instance.state.showRelativeDate).toBe(false);
    dom.root.instance.toggleRelativeDate();
    expect(dom.root.instance.state.showRelativeDate).toBe(true);
  });

  it("kills timer on unmount", () => {
    const dom = create(<RelativeDate date={date} />);
    const timer = dom.root.instance.timer;
    expect(timer._onTimeout).not.toBeNull();
    dom.unmount();
    expect(timer._onTimeout).toBeNull();
  });
});
