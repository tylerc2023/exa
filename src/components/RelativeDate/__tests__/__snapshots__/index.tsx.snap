// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 14,
          "fontWeight": "300",
          "letterSpacing": 0.25,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    a few seconds ago
  </Text>
</View>
`;

exports[`renders correctly on touch 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 14,
          "fontWeight": "300",
          "letterSpacing": 0.25,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Mar 29 2019, 5:11 pm
  </Text>
</View>
`;
